File: backend/app/extraction/ifc_geometry.py
"""IFC solid-geometry helpers: turns an IfcPavement's triangulated shell into
width-vs-station samples, per the validated method — project vertices onto the
axis, group them into transversal 'rings' by station proximity, and take each
ring's lateral (perpendicular-to-axis) extent as the width at that PK. Using
the offset span rather than pairing two specific vertices makes this robust
to rings carrying more than two vertices (e.g. a solid with wall thickness
contributing duplicate top/bottom vertices at the same plan position)."""

from __future__ import annotations

import numpy as np
import ifcopenshell
import ifcopenshell.geom

from app.extraction.axis_reference import AxisReference

_SETTINGS = ifcopenshell.geom.settings()
_SETTINGS.set(_SETTINGS.USE_WORLD_COORDS, True)


def shape_vertices(product) -> np.ndarray | None:
    try:
        shape = ifcopenshell.geom.create_shape(_SETTINGS, product)
    except Exception:
        return None
    verts = shape.geometry.verts
    if not verts:
        return None
    arr = np.array(verts, dtype=float).reshape(-1, 3)
    return arr


def all_vertices(
    ifc,
    ifc_class: str = "IfcPavement",
    max_products: int = 500,
    max_vertices: int = 200_000,
) -> np.ndarray:
    """Collects vertices for the PCA fallback axis. Bounded on purpose: this
    only needs enough points to fit a stable direction and a binned
    centerline, not the whole model. Without a cap, a reference file that
    happens to carry no IfcPavement falls through to triangulating *every*
    IfcProduct in the file — on a real full BIM export (as opposed to the
    small synthetic fixtures this was tested against) that is enough
    triangulated geometry to exceed a 512 MB container and get OOM-killed."""
    chunks = []
    total = 0
    for product in ifc.by_type(ifc_class):
        v = shape_vertices(product)
        if v is not None:
            chunks.append(v)
            total += len(v)
        if total >= max_vertices:
            break
    if not chunks:
        count = 0
        for product in ifc.by_type("IfcProduct"):
            v = shape_vertices(product)
            if v is not None:
                chunks.append(v)
                total += len(v)
                count += 1
            if count >= max_products or total >= max_vertices:
                break
    if not chunks:
        raise ValueError("No geometry found in IFC file to build a fallback axis")
    return np.concatenate(chunks, axis=0)


def pca_axis_polyline(verts: np.ndarray, n_bins: int = 40) -> list[tuple[float, float]]:
    """Fallback axis when no IfcAlignment is present: PCA principal direction
    of the point cloud, then bin points along it and take per-bin centroids so
    the resulting polyline follows the corridor's curvature rather than being
    a single straight chord."""
    xy = verts[:, :2]
    centroid = xy.mean(axis=0)
    centered = xy - centroid
    _, _, vt = np.linalg.svd(centered, full_matrices=False)
    direction = vt[0]
    projection = centered @ direction
    order = np.argsort(projection)
    proj_sorted = projection[order]
    bins = np.linspace(proj_sorted[0], proj_sorted[-1], n_bins + 1)
    bin_idx = np.clip(np.digitize(projection, bins) - 1, 0, n_bins - 1)
    pts = []
    for b in range(n_bins):
        mask = bin_idx == b
        if mask.any():
            pts.append(tuple(xy[mask].mean(axis=0)))
    if len(pts) < 2:
        raise ValueError("Not enough spread in IFC geometry to derive a fallback axis")
    return pts


def cluster_ring_widths(
    stations: np.ndarray,
    offsets: np.ndarray,
    gap_threshold_m: float = 1.5,
    plausible_range: tuple[float, float] = (0.1, 20.0),
) -> list[tuple[float, float]]:
    """Pure clustering logic, kept separate from IFC shape extraction so it can
    be unit-tested with synthetic (station, offset) pairs. Sorts by station,
    greedily chains vertices into the same ring while consecutive stations are
    closer than `gap_threshold_m`, then takes each ring's offset span (max -
    min) as its width. Rings whose span falls outside `plausible_range` are
    dropped (rejects near-zero noise and cross-ring mismatches)."""
    order = np.argsort(stations, kind="stable")
    stations_sorted = stations[order]
    offsets_sorted = offsets[order]

    groups: list[list[int]] = []
    current = [0]
    for i in range(1, len(stations_sorted)):
        if stations_sorted[i] - stations_sorted[current[-1]] < gap_threshold_m:
            current.append(i)
        else:
            groups.append(current)
            current = [i]
    groups.append(current)

    samples: list[tuple[float, float]] = []
    for group in groups:
        if len(group) < 2:
            continue
        group_offsets = offsets_sorted[group]
        group_stations = stations_sorted[group]
        width = float(group_offsets.max() - group_offsets.min())
        if plausible_range[0] <= width <= plausible_range[1]:
            samples.append((float(group_stations.mean()), width))
    return samples


def pavement_width_samples(
    product,
    axis: AxisReference,
    gap_threshold_m: float = 1.5,
    plausible_range: tuple[float, float] = (0.1, 20.0),
) -> list[tuple[float, float]]:
    """Returns a list of (pk, width_m) for one IfcPavement solid."""
    verts = shape_vertices(product)
    if verts is None or len(verts) < 2:
        return []
    projections = np.array([axis.axis.project_point((v[0], v[1])) for v in verts])
    stations, offsets = projections[:, 0], projections[:, 1]
    pairs = cluster_ring_widths(stations, offsets, gap_threshold_m, plausible_range)
    return [(axis.station_to_pk(station), width) for station, width in pairs]

File: frontend/src/engine/ifc/ifcGeometry.ts
/** Port of backend/app/extraction/ifc_geometry.py. */

import type { IfcAPI } from "web-ifc";
import type { AxisReference } from "../axisReference";
import { perpendicularDirection, type Point } from "../geometry";
import { shapeVertices } from "./webIfcClient";
import { maxOf, minOf, pushAll } from "../arrayUtils";

export interface RingWidth {
  station: number;
  width: number;
  /** Signed axis offsets bounding this ring — offsetNear has the smaller
   * absolute value (closer to the axis), offsetFar the larger. Kept
   * alongside width so callers can reconstruct the true plan-view boundary
   * points of this ring, not just its scalar width. */
  offsetNear: number;
  offsetFar: number;
}

export function clusterRingWidths(
  stations: number[],
  offsets: number[],
  gapThresholdM = 1.5,
  plausibleRange: [number, number] = [0.1, 20.0],
): RingWidth[] {
  const order = stations.map((_, i) => i).sort((a, b) => stations[a] - stations[b]);
  const stationsSorted = order.map((i) => stations[i]);
  const offsetsSorted = order.map((i) => offsets[i]);

  const groups: number[][] = [];
  let current = [0];
  for (let i = 1; i < stationsSorted.length; i++) {
    if (stationsSorted[i] - stationsSorted[current[current.length - 1]] < gapThresholdM) {
      current.push(i);
    } else {
      groups.push(current);
      current = [i];
    }
  }
  groups.push(current);

  const samples: RingWidth[] = [];
  for (const group of groups) {
    if (group.length < 2) continue;
    const groupOffsets = group.map((i) => offsetsSorted[i]);
    const groupStations = group.map((i) => stationsSorted[i]);
    const offsetMin = minOf(groupOffsets);
    const offsetMax = maxOf(groupOffsets);
    const width = offsetMax - offsetMin;
    if (width >= plausibleRange[0] && width <= plausibleRange[1]) {
      const meanStation = groupStations.reduce((a, b) => a + b, 0) / groupStations.length;
      const [offsetNear, offsetFar] = Math.abs(offsetMin) <= Math.abs(offsetMax) ? [offsetMin, offsetMax] : [offsetMax, offsetMin];
      samples.push({ station: meanStation, width, offsetNear, offsetFar });
    }
  }
  return samples;
}

/** Fallback axis when no IfcAlignment is present: PCA principal direction of
 * the point cloud, then bin points along it and take per-bin centroids. */
export function pcaAxisPolyline(verts: [number, number, number][], nBins = 40): Point[] {
  const xy: Point[] = verts.map((v) => [v[0], v[1]]);
  const cx = xy.reduce((s, p) => s + p[0], 0) / xy.length;
  const cy = xy.reduce((s, p) => s + p[1], 0) / xy.length;
  const centered = xy.map(([x, y]) => [x - cx, y - cy] as Point);

  // Principal direction via the 2x2 covariance matrix's dominant eigenvector
  // (equivalent to the first right-singular vector used in the Python SVD).
  let sxx = 0;
  let sxy = 0;
  let syy = 0;
  for (const [x, y] of centered) {
    sxx += x * x;
    sxy += x * y;
    syy += y * y;
  }
  const trace = sxx + syy;
  const det = sxx * syy - sxy * sxy;
  const eig = trace / 2 + Math.sqrt(Math.max(0, (trace * trace) / 4 - det));
  let dx = sxy;
  let dy = eig - sxx;
  if (Math.abs(dx) < 1e-12 && Math.abs(dy) < 1e-12) {
    dx = 1;
    dy = 0;
  }
  const norm = Math.hypot(dx, dy);
  dx /= norm;
  dy /= norm;

  const projection = centered.map(([x, y]) => x * dx + y * dy);
  const minP = minOf(projection);
  const maxP = maxOf(projection);
  const binWidth = (maxP - minP) / nBins;
  const bins: Point[][] = Array.from({ length: nBins }, () => []);
  for (let i = 0; i < projection.length; i++) {
    let b = binWidth > 1e-12 ? Math.floor((projection[i] - minP) / binWidth) : 0;
    b = Math.min(Math.max(b, 0), nBins - 1);
    bins[b].push(xy[i]);
  }
  const pts: Point[] = [];
  for (const bin of bins) {
    if (bin.length === 0) continue;
    pts.push([bin.reduce((s, p) => s + p[0], 0) / bin.length, bin.reduce((s, p) => s + p[1], 0) / bin.length]);
  }
  if (pts.length < 2) throw new Error("Not enough spread in IFC geometry to derive a fallback axis");
  return pts;
}

/** Bounded on purpose — see backend/app/extraction/ifc_geometry.py's
 * all_vertices docstring: without a cap, a reference file with no
 * IfcPavement falls through to triangulating every IfcProduct in the model. */
export function allVertices(
  api: IfcAPI,
  modelID: number,
  pavementExpressIds: number[],
  otherProductExpressIds: () => Iterable<number>,
  maxProducts = 500,
  maxVertices = 200_000,
): [number, number, number][] {
  const chunks: [number, number, number][] = [];
  let total = 0;
  for (const id of pavementExpressIds) {
    const v = shapeVertices(api, modelID, id);
    if (v) {
      pushAll(chunks, v);
      total += v.length;
    }
    if (total >= maxVertices) break;
  }
  if (chunks.length === 0) {
    let count = 0;
    for (const id of otherProductExpressIds()) {
      const v = shapeVertices(api, modelID, id);
      if (v) {
        pushAll(chunks, v);
        total += v.length;
        count++;
      }
      if (count >= maxProducts || total >= maxVertices) break;
    }
  }
  if (chunks.length === 0) throw new Error("No geometry found in IFC file to build a fallback axis");
  return chunks;
}

export interface PlanWidthSample {
  pk: number;
  width: number;
  /** True (x, y) boundary points of this ring, reconstructed from the axis
   * at this station — used to redraw the road in plan (see dxfExport.ts)
   * instead of only a schematic (pk, width) chart. */
  near: Point;
  far: Point;
}

export function pavementWidthSamples(
  api: IfcAPI,
  modelID: number,
  expressID: number,
  axis: AxisReference,
  gapThresholdM = 1.5,
  plausibleRange: [number, number] = [0.1, 20.0],
): PlanWidthSample[] {
  const verts = shapeVertices(api, modelID, expressID);
  if (!verts || verts.length < 2) return [];
  const stations: number[] = [];
  const offsets: number[] = [];
  for (const v of verts) {
    const [s, o] = axis.axis.projectPoint([v[0], v[1]]);
    stations.push(s);
    offsets.push(o);
  }
  const rings = clusterRingWidths(stations, offsets, gapThresholdM, plausibleRange);
  return rings.map(({ station, width, offsetNear, offsetFar }) => {
    const { point, direction } = axis.axis.pointAndDirectionAtStation(station);
    const perp = perpendicularDirection(direction);
    const near: Point = [point[0] + offsetNear * perp[0], point[1] + offsetNear * perp[1]];
    const far: Point = [point[0] + offsetFar * perp[0], point[1] + offsetFar * perp[1]];
    return { pk: axis.stationToPk(station), width, near, far };
  });
}
